yamux/src/connection/rtt.rs

```diff
@@ -0,0 +1,170 @@
+// at https://www.apache.org/licenses/LICENSE-2.0 and a copy of the MIT license
+// at https://opensource.org/licenses/MIT.
+
+//! Connection round-trip time measurement
+
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use parking_lot::Mutex;
+
+use crate::connection::Action;
+use crate::frame::{header::Ping, Frame};
+
+const PING_INTERVAL: Duration = Duration::from_secs(10);
+
+#[derive(Clone, Debug)]
+pub(crate) struct Rtt(Arc<Mutex<RttInner>>);
+
+impl Rtt {
+    pub(crate) fn new() -> Self {
+        Self(Arc::new(Mutex::new(RttInner {
+            rtt: None,
+            state: RttState::Initial,
+        })))
+    }
+
+    pub(crate) fn next_ping(&mut self) -> Option<Frame<Ping>> {
+        let state = &mut self.0.lock().state;
+
+        match state {
+            RttState::Initial => {}
+            RttState::AwaitingPong { .. } => return None,
+            RttState::Waiting { next } => {
+                if *next > Instant::now() {
+                    return None;
+                }
+            }
+        }
+
+        let nonce = rand::random();
+        *state = RttState::AwaitingPong {
+            sent_at: Instant::now(),
+            nonce,
+        };
+        log::debug!("sending ping {nonce}");
+        Some(Frame::ping(nonce))
+    }
+
+    pub(crate) fn handle_pong(&mut self, received_nonce: u32) -> Action {
+        let inner = &mut self.0.lock();
+
+        let (sent_at, expected_nonce) = match inner.state {
+            RttState::Initial | RttState::Waiting { .. } => {
+                log::error!("received unexpected pong {}", received_nonce);
+                return Action::Terminate(Frame::protocol_error());
+            }
+            RttState::AwaitingPong { sent_at, nonce } => (sent_at, nonce),
+        };
+
+        if received_nonce != expected_nonce {
+            log::error!(
+                "received pong with {} but expected {}",
+                received_nonce,
+                expected_nonce
+            );
+            return Action::Terminate(Frame::protocol_error());
+        }
+
+        inner.rtt = Some(sent_at.elapsed());
+        log::debug!(
+            "received pong {received_nonce}, estimated round-trip-time {}s",
+            inner.rtt.as_ref().unwrap().as_secs_f64()
+        );
+
+        inner.state = RttState::Waiting {
+            next: Instant::now() + PING_INTERVAL,
+        };
+
+        return Action::None;
+    }
+
+    pub(crate) fn get(&self) -> Option<Duration> {
+        self.0.lock().rtt
+    }
+}
+
+#[cfg(test)]
+impl quickcheck::Arbitrary for Rtt {
+    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
+        Self(Arc::new(Mutex::new(RttInner::arbitrary(g))))
+    }
+}
+
+#[derive(Debug)]
+struct RttInner {
+    state: RttState,
+    rtt: Option<Duration>,
+}
+
+#[cfg(test)]
+impl Clone for RttInner {
+    fn clone(&self) -> Self {
+        Self {
+            state: self.state.clone(),
+            rtt: self.rtt.clone(),
+        }
+    }
+}
```

Comment:
Could be expressed as:

```rust
#[cfg_attr(test, derive(Clone))]
```

Reply:
Oh neat! Thanks @thomaseizinger!

82106b0